bsondump/bsondump.go

```diff
@@ -79,27 +80,58 @@ func (bdo *BSONDumpOptions) GetBSONReader() (io.ReadCloser, error) {
 	return ReadNopCloser{os.Stdin}, nil
 }
 
-func formatJSON(doc *bson.Raw, pretty bool) ([]byte, error) {
-	decodedDoc := bson.D{}
-	err := bson.Unmarshal(doc.Data, &decodedDoc)
+// New constructs a new instance of BSONDump configured by the provided options.
+// A successfully created instance must be closed with Close().
+func New(opts Options) (*BSONDump, error) {
+	dumper := &BSONDump{
+		ToolOptions:   opts.ToolOptions,
+		OutputOptions: opts.OutputOptions,
+	}
+
+	reader, err := opts.GetBSONReader()
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("getting BSON reader failed: %v", err)
+	}
+	dumper.BSONSource = db.NewBSONSource(reader)
+
+	writer, err := opts.GetWriter()
+	if err != nil {
+		_ = reader.Close()
+		return nil, fmt.Errorf("getting Writer failed: %v", err)
+	}
+	dumper.Out = writer
+
+	return dumper, nil
+}
+
+// Close cleans up the internal state of the given BSONDump instance. The instance should not be used again
+// after Close is called.
+func (bd *BSONDump) Close() error {
+	if err := bd.Out.Close(); err != nil {
+		return err
+	}
+
+	if err := bd.BSONSource.Close(); err != nil {
```

Comment:
I have it fail early this way since, if a call to `Close` fails, the caller shouldn't expect all of the resources to have been cleaned up, so we might as well not even try. Also, realistically regardless of the outcome of `Close`, the program will just exit anyhow. That said, I don't really know what the best answer is, this is just my reasoning was.

Reply:
that makes sense. I'm ok with leaving it as is.